homeassistant/components/growatt_server/coordinator.py

```diff
@@ -0,0 +1,215 @@
+        self.api.server_url = self.url
+
+        self.device_id = device_id
+        self.device_type = device_type
+        self.plant_id = plant_id
+        self.data: dict[str, Any] = {}
+        self.previous_values: dict[str, Any] = {}
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN} ({device_id})",
+            update_interval=SCAN_INTERVAL,
+        )
+
+    async def _async_update_data(self):
+        """Update data via library."""
+        try:
+            _LOGGER.debug("Updating data for %s (%s)", self.device_id, self.device_type)
+
+            # Login in to the Growatt server
+            await self.hass.async_add_executor_job(
+                self.api.login, self.username, self.password
+            )
+
+            if self.device_type == "total":
+                total_info = await self.hass.async_add_executor_job(
+                    self.api.plant_info, self.device_id
+                )
+                del total_info["deviceList"]
+                plant_money_text, currency = total_info["plantMoneyText"].split("/")
+                total_info["plantMoneyText"] = plant_money_text
+                total_info["currency"] = currency
+                self.data = total_info
+            elif self.device_type == "inverter":
+                inverter_info = await self.hass.async_add_executor_job(
+                    self.api.inverter_detail, self.device_id
+                )
+                self.data = inverter_info
+            elif self.device_type == "tlx":
+                tlx_info = await self.hass.async_add_executor_job(
+                    self.api.tlx_detail, self.device_id
+                )
+                self.data = tlx_info["data"]
+            elif self.device_type == "storage":
+                storage_info_detail = await self.hass.async_add_executor_job(
+                    self.api.storage_params, self.device_id
+                )
+                storage_energy_overview = await self.hass.async_add_executor_job(
+                    self.api.storage_energy_overview, self.plant_id, self.device_id
+                )
+                self.data = {
+                    **storage_info_detail["storageDetailBean"],
+                    **storage_energy_overview,
+                }
+            elif self.device_type == "mix":
+                mix_info = await self.hass.async_add_executor_job(
+                    self.api.mix_info, self.device_id
+                )
+                mix_totals = await self.hass.async_add_executor_job(
+                    self.api.mix_totals, self.device_id, self.plant_id
+                )
+                mix_system_status = await self.hass.async_add_executor_job(
+                    self.api.mix_system_status, self.device_id, self.plant_id
+                )
+                mix_detail = await self.hass.async_add_executor_job(
+                    self.api.mix_detail, self.device_id, self.plant_id
+                )
```

Comment:
So, switching from the event loop to an executor job is expensive, so we try to avoid running multiple after each other. So I think the easiest for now (to make the code change clear), is to just move the content of the current `_async_update_data` and move it to a sync function and then make `_async_update_data` run that new function in the executor job. This way we are more efficient